File: fetch_courses.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
同济 1.tongji.edu.cn EnquiryOfCourses 选课数据抓取（按页面筛选条件）

用法：
    python fetch_courses.py options    # 打印所有可选真实选项（学期/开课学院/课程性质/校区/培养层次/学习形式）
    python fetch_courses.py            # 按下方 FILTERS 拉数据 -> JSON + CSV

鉴权：
    通过环境变量 ONESYSTEM_X_TOKEN 提供 sessionStorage 中的 sessionid；
    可选用 ONESYSTEM_COOKIE 提供整段 Cookie。不要把密钥写入源码。
"""

import json
import time
import sys
import os
import csv
import datetime
import re
import urllib.request
import urllib.error

# ======================= 需要你改的部分 =======================

X_TOKEN = os.environ.get("ONESYSTEM_X_TOKEN", "")
COOKIE = os.environ.get("ONESYSTEM_COOKIE", "")

# 筛选条件：全部留空 = 拉全部。只填你想筛的项，其余留空字符串。
# 每个选项的"真实值"可用  python fetch_courses.py options  查出来再回填。
FILTERS = {
    # 学期/学年（必填其一）：'current'=当前学期；或写 options 打印的学期 id，如 '122'
    "calendarId": "current",

    # 课程编码（页面“课程编码”输入框，对应接口 newCourseCode）
    "newCourseCode": "",
    # 课程代码/教学班号（页面可选显示的 courseCode，对应接口 teachClassCode）
    "teachClassCode": "",
    # 课程名称（模糊匹配）
    "courseName": "",
    # 教师姓名（模糊匹配）
    "teacherName": "",
    # 开课学院（填 deptCode，如研究生院='000014'，数学科学学院='...'，用 options 查）
    "faculty": "",
    # 课程性质（X_KCXZ：'1'公共课 / '2'专业课 / '3'必修环节）
    "nature": "",
    # 校区（X_XQ：'1'四平路 / '2'沪北 / '3'嘉定 / '4'沪西 / '5'其他）
    "campu": "",
    # 培养层次（X_PYCC：'4'硕士 / '6'博士）
    "trainingLevel": "",
    # 学习形式（K_XXXS：'1'全日制 / '2'非全日制 / '4'全日制&非全日制）
    "formLearning": "",
}

# ======================= 以下一般不用改 =======================

BASE = "https://1.tongji.edu.cn"
PAGE_SIZE = 1000
DELAY_MS = 0.2
OUT_DIR = "."

# 页面 queryForm 的键名 -> 本脚本 FILTERS 键名（campus 在接口里拼作 campu，照抄页面）
_API_COND = {
    "calendarId": "calendarId",
    "newCourseCode": "newCourseCode",
    "teachClassCode": "teachClassCode",
    "courseName": "courseName",
    "teacherName": "teacherName",
    "faculty": "faculty",
    "nature": "nature",
    "campu": "campu",
    "trainingLevel": "trainingLevel",
    "formLearning": "formLearning",
}
_DICT_KEYS = ["X_KCXZ", "X_XQ", "X_PYCC", "K_XXXS"]
_DICT_LABEL = {"X_KCXZ": "课程性质(nature)", "X_XQ": "校区(campu)",
               "X_PYCC": "培养层次(trainingLevel)", "K_XXXS": "学习形式(formLearning)"}
_SAFE_FIELDS = {
    "calendarId", "id", "code", "name", "courseLabelId", "courseLabelName",
    "assessmentMode", "assessmentModeI18n", "period", "weekHour", "campus",
    "campusI18n", "number", "elcNumber", "startWeek", "endWeek", "courseCode",
    "courseName", "credits", "credit", "teachingLanguage", "teachingLanguageI18n",
    "faculty", "facultyI18n", "newCourseCode", "newCode", "arrangeInfo",
    "teacherList", "majorList", "timeTableList", "trainingLevel", "formLearning",
    "nature", "campu", "teachClassCode", "capacity", "selectedNumber", "totalNumber",
}
_PRIVATE_FIELD_PARTS = ("student", "selection", "contact", "phone", "mobile", "email",
                       "identity", "idcard", "password", "token", "cookie")


def http_json(path, payload=None):
    headers = {
        "X-Token": X_TOKEN,
        "Accept": "application/json, text/plain, */*",
        "Referer": BASE + "/EnquiryOfCourses",
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
    }
    if COOKIE:
        headers["Cookie"] = COOKIE
    data = None
    if payload is not None:
        headers["Content-Type"] = "application/json;charset=UTF-8"
        data = json.dumps(payload, ensure_ascii=False).encode("utf-8")
    req = urllib.request.Request(BASE + path, data=data, headers=headers, method="POST" if payload is not None else "GET")
    try:
        with urllib.request.urlopen(req, timeout=60) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "ignore")
        if e.code == 401 or "sessionid is not exist" in body:
            raise RuntimeError("401 未授权：X-Token 无效或已过期，请重新获取 sessionid") from None
        raise RuntimeError("HTTP %s" % e.code) from None


# ----------------------- 选项查询 -----------------------

def current_term():
    j = http_json("/api/baseresservice/schoolCalendar/currentTermCalendar?flag=0")
    d = (j.get("data") or {}).get("schoolCalendar") or j.get("data") or {}
    return str(d.get("id")), d.get("simpleName") or d.get("name") or d.get("id")


def list_calendars():
    j = http_json("/api/baseresservice/schoolCalendar/list")
    out = []
    for c in j.get("data") or []:
        out.append((str(c["id"]), c.get("fullName") or str(c.get("id"))))
    return out


def list_faculties():
    j = http_json("/api/electionservice/elcMutualCourses/findDept?manageDept=0&type=1&virtualDept=0")
    items = j.get("data") or []
    if isinstance(items, dict):
        items = list(items.values())
    out = []
    for it in items:
        out.append((str(it.get("deptCode")), it.get("deptName")))
    out.sort(key=lambda x: x[1])
    return out


def list_dicts():
    j = http_json("/api/commonservice/dictionary/query",
                  {"lang": "cn", "type": "allChild", "keys": _DICT_KEYS})
    return j.get("data") or {}


def cmd_options():
    cur_id, cur_name = current_term()
    print("== 当前学期 ==")
    print("  %s  %s\n" % (cur_id, cur_name))
    print("== 全部学期（FILTERS['calendarId'] 可填 id 或 'current'）==")
    for cid, name in list_calendars():
        mark = "  <- 当前" if cid == cur_id else ""
        print("  %s  %s%s" % (cid, name, mark))
    print("\n== 开课学院（FILTERS['faculty'] 填第一列 deptCode）==")
    for code, name in list_faculties():
        print("  %s  %s" % (code, name))
    print("\n== 字典选项（value 填进对应 FILTERS 项）==")
    dicts = list_dicts()
    for dk, label in _DICT_LABEL.items():
        print("  -- %s --" % label)
        for code, name in (dicts.get(dk) or {}).items():
            print("    %s  %s" % (code, name))


# ----------------------- 数据抓取 -----------------------

def fetch_all(condition):
    rows, page, total = [], 1, None
    while True:
        body = {
            "pageNum_": page,
            "pageSize_": PAGE_SIZE,
            "condition": condition,
        }
        j = http_json("/api/electionservice/student/round/allArrangementCourses", body)
        data = j.get("data") or {}
        lst = data.get("list") or []
        rows.extend(lst)
        total = data.get("total_", data.get("total"))
        print("\r  第 %d 页 +%d 条（累计 %d%s）    " % (page, len(lst), len(rows), (" / " + str(total)) if total is not None else ""), end="")
        if not lst:
            break
        if total is not None and len(rows) >= int(total):
            break
        if len(lst) < PAGE_SIZE:
            break
        page += 1
        time.sleep(DELAY_MS)
    print()
    return rows


def build_condition():
    cond = {}
    for fk, api_key in _API_COND.items():
        val = str(FILTERS.get(fk) or "").strip()
        if fk == "calendarId":
            if not val or val == "current":
                cid, _ = current_term()
                cond[api_key] = cid
            else:
                cond[api_key] = val
        else:
            cond[api_key] = val  # 与网页一致：空串=不过滤
    return cond


def sanitize_value(value):
    if isinstance(value, list):
        return [sanitize_value(item) for item in value]
    if isinstance(value, dict):
        out = {}
        for key, item in value.items():
            if any(part in str(key).lower() for part in _PRIVATE_FIELD_PARTS):
                raise RuntimeError("响应包含禁止导出的隐私字段: %s" % key)
            out[key] = sanitize_value(item)
        return out
    return value


def sanitize_rows(rows):
    safe_rows = []
    for row in rows:
        safe = {}
        for key, value in row.items():
            if any(part in str(key).lower() for part in _PRIVATE_FIELD_PARTS):
                raise RuntimeError("响应包含禁止导出的隐私字段: %s" % key)
            if key in _SAFE_FIELDS:
                safe[key] = sanitize_value(value)
        safe_rows.append(safe)
    return safe_rows


def csv_value(value):
    if isinstance(value, (dict, list)):
        value = json.dumps(value, ensure_ascii=False)
    text = "" if value is None else str(value)
    if text[:1] in ("=", "+", "-", "@", "\t", "\r", "\n"):
        return "'" + text
    return text


def write_outputs(rows, cal_id, cond):
    ts = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
    base = os.path.join(OUT_DIR, "courses-%s-%s" % (cal_id, ts))
    rows = sanitize_rows(rows)
    with open(base + "-sanitized.json", "w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False, indent=1)
    keys = []
    for r in rows:
        for k in r.keys():
            if k not in keys:
                keys.append(k)
    with open(base + ".csv", "w", encoding="utf-8-sig", newline="") as f:
        w = csv.writer(f)
        w.writerow(keys)
        for r in rows:
            w.writerow([csv_value(r.get(k)) for k in keys])
    summary = "; ".join("%s=%s" % (k, v) for k, v in cond.items() if v)
    print("完成：%d 条  筛选[%s]" % (len(rows), summary))
    print("输出: %s-sanitized.json / %s.csv" % (base, base))


def main():
    token = X_TOKEN.strip()
    if not token or token.startswith("在此粘贴") or re.fullmatch(r"x{8,}", token, re.IGNORECASE):
        raise SystemExit("请通过 ONESYSTEM_X_TOKEN 环境变量提供 sessionid（不要写入源码）")
    if len(sys.argv) > 1 and sys.argv[1] == "options":
        cmd_options()
        return
    cond = build_condition()
    print("抓取条件:", cond)
    rows = fetch_all(cond)
    write_outputs(rows, cond.get("calendarId"), cond)


if __name__ == "__main__":
    main()

File: fetch_courses.js
/* ============================================================
 * 同济 1.tongji.edu.cn EnquiryOfCourses 选课信息抓取（Node 版）
 *
 * 用法：
 *   node fetch_courses.js options    # 打印所有可选真实选项（学期/开课学院/课程性质/校区/培养层次/学习形式）
 *   node fetch_courses.js            # 按下方 FILTERS 抓取 -> JSON + CSV
 *
 * 鉴权：通过环境变量 ONESYSTEM_X_TOKEN 提供 sessionStorage 中的 sessionid；
 *       可选用 ONESYSTEM_COOKIE 提供整段 Cookie。不要把密钥写入源码。
 * ============================================================ */
'use strict';

const fs = require('fs');

const CONFIG = {
  HOST: 'https://1.tongji.edu.cn',

  X_TOKEN: process.env.ONESYSTEM_X_TOKEN || '',

  /* 可选：整段 Cookie；一般有 X_TOKEN 即可，留空也行。 */
  COOKIE: process.env.ONESYSTEM_COOKIE || '',

  /* 筛选条件：全部留空 = 拉全部。真实选项值用 node fetch_courses.js options 查询后回填。 */
  FILTERS: {
    calendarId: 'current',   // 学期：'current'=当前学期，或填 options 打印的 id，如 '122'
    newCourseCode: '',       // 课程编码
    teachClassCode: '',      // 课程代码/教学班号
    courseName: '',          // 课程名称（模糊）
    teacherName: '',         // 教师姓名（模糊）
    faculty: '',             // 开课学院 deptCode，如 '000014'（研究生院）
    nature: '',              // 课程性质：1公共课 2专业课 3必修环节
    campu: '',               // 校区：1四平路 2沪北 3嘉定 4沪西 5其他（接口拼写即 campu）
    trainingLevel: '',       // 培养层次：4硕士 6博士
    formLearning: '',        // 学习形式：1全日制 2非全日制 4全日制&非全日制
  },

  PAGE_SIZE: 1000,   // 单页条数（接口上限 1000）
  DELAY_MS: 200,     // 每页间隔，避免压测
};

/* ---------------- 请求封装 ---------------- */

async function api(pathname, opts = {}) {
  const headers = {
    'X-Token': CONFIG.X_TOKEN,
    'X-Requested-With': 'XMLHttpRequest',
    'Accept': 'application/json, text/plain, */*',
    'Referer': CONFIG.HOST + '/EnquiryOfCourses',
  };
  if (CONFIG.COOKIE) headers['Cookie'] = CONFIG.COOKIE;
  if (opts.body) headers['Content-Type'] = 'application/json;charset=UTF-8';

  const res = await fetch(CONFIG.HOST + pathname, {
    method: opts.method || (opts.body ? 'POST' : 'GET'),
    headers,
    body: opts.body ? JSON.stringify(opts.body) : undefined,
  });
  const text = await res.text();
  let json;
  try { json = JSON.parse(text); } catch (e) { throw new Error('非 JSON 响应 HTTP ' + res.status); }
  if (res.status === 401 || /sessionid is not exist/i.test(text)) {
    throw new Error('401 未授权：X-Token 无效或已过期，请重新获取 sessionid');
  }
  if (json.code != null && json.code !== 200) {
    throw new Error('接口返回 code=' + json.code);
  }
  return json;
}

/* ---------------- 选项（学期 / 学院 / 字典） ---------------- */

async function fetchCalendars() {
  const j = await api('/api/baseresservice/schoolCalendar/list');
  return (j.data || []).map(c => ({ id: String(c.id), name: c.fullName || String(c.id) }));
}

async function currentTerm() {
  const j = await api('/api/baseresservice/schoolCalendar/currentTermCalendar?flag=0');
  const d = (j.data && j.data.schoolCalendar) || j.data || {};
  return { id: String(d.id), name: d.simpleName || d.name || String(d.id) };
}

async function fetchFaculties() {
  const j = await api('/api/electionservice/elcMutualCourses/findDept?manageDept=0&type=1&virtualDept=0');
  const data = Array.isArray(j.data) ? j.data : Object.values(j.data || {});
  return data.map(x => ({ code: String(x.deptCode), name: x.deptName }));
}

async function fetchDicts() {
  const j = await api('/api/commonservice/dictionary/query', {
    body: { lang: 'cn', type: 'allChild', keys: ['X_KCXZ', 'X_XQ', 'X_PYCC', 'K_XXXS'] },
  });
  return j.data || {};
}

async function cmdOptions() {
  const cur = await currentTerm();
  console.log('== 当前学期 ==\n  ' + cur.id + '  ' + cur.name + '\n');
  console.log('== 全部学期（FILTERS.calendarId 填 id 或 current）==');
  for (const c of await fetchCalendars()) console.log('  ' + c.id + (c.id === cur.id ? '  <- 当前' : '') + '  ' + c.name);
  console.log('\n== 开课学院（FILTERS.faculty 填 deptCode）==');
  for (const f of await fetchFaculties()) console.log('  ' + f.code + '  ' + f.name);
  const label = { X_KCXZ: '课程性质 nature', X_XQ: '校区 campu', X_PYCC: '培养层次 trainingLevel', K_XXXS: '学习形式 formLearning' };
  const dicts = await fetchDicts();
  for (const k of Object.keys(label)) {
    console.log('\n-- ' + label[k] + ' --');
    for (const [code, name] of Object.entries(dicts[k] || {})) console.log('    ' + code + '  ' + name);
  }
}

/* ---------------- 抓取 ---------------- */

async function fetchAll(calendarId) {
  const rows = [];
  let page = 1;
  let total = null;
  for (;;) {
    const condition = { ...CONFIG.FILTERS, calendarId: String(calendarId) };
    const body = { pageNum_: page, pageSize_: CONFIG.PAGE_SIZE, condition };
    const j = await api('/api/electionservice/student/round/allArrangementCourses', { body });
    const data = j.data || {};
    const list = Array.isArray(data.list) ? data.list : [];
    rows.push(...list);
    total = data.total_ != null ? data.total_ : (data.total != null ? data.total : null);
    process.stdout.write(`\r  第 ${page} 页 +${list.length} 条（累计 ${rows.length}${total != null ? ' / ' + total : ''}）   `);
    if (!list.length) break;
    if (total != null && rows.length >= total) break;
    if (list.length < CONFIG.PAGE_SIZE) break;
    page++;
    await new Promise(r => setTimeout(r, CONFIG.DELAY_MS));
  }
  process.stdout.write('\n');
  return rows;
}

/* ---------------- 输出 ---------------- */

// 只导出课程查询所需的非个人字段；上游新增字段默认不落盘，避免快照意外
// 扩大为选课关系/联系方式等个人数据。
const SAFE_FIELDS = new Set([
  'calendarId', 'id', 'code', 'name', 'courseLabelId', 'courseLabelName',
  'assessmentMode', 'assessmentModeI18n', 'period', 'weekHour', 'campus',
  'campusI18n', 'number', 'elcNumber', 'startWeek', 'endWeek', 'courseCode',
  'courseName', 'credits', 'credit', 'teachingLanguage', 'teachingLanguageI18n',
  'faculty', 'facultyI18n', 'newCourseCode', 'newCode', 'arrangeInfo',
  'teacherList', 'majorList', 'timeTableList', 'trainingLevel', 'formLearning',
  'nature', 'campu', 'teachClassCode', 'capacity', 'selectedNumber', 'totalNumber',
]);
const PRIVATE_FIELD = /(?:student|selection|selectedStudent|contact|phone|mobile|email|identity|idCard|password|token|cookie)/i;

function sanitizeValue(value, fieldName) {
  if (Array.isArray(value)) return value.map(item => sanitizeValue(item, fieldName));
  if (!value || typeof value !== 'object') return value;
  const out = {};
  for (const [key, item] of Object.entries(value)) {
    if (PRIVATE_FIELD.test(key)) throw new Error('响应包含禁止导出的隐私字段: ' + key);
    out[key] = sanitizeValue(item, key);
  }
  return out;
}

function sanitizeRows(rows) {
  return rows.map(row => {
    const out = {};
    for (const [key, value] of Object.entries(row)) {
      if (PRIVATE_FIELD.test(key)) throw new Error('响应包含禁止导出的隐私字段: ' + key);
      if (SAFE_FIELDS.has(key)) out[key] = sanitizeValue(value, key);
    }
    return out;
  });
}

function toCSV(rows) {
  const keys = [];
  for (const r of rows) for (const k of Object.keys(r)) if (!keys.includes(k)) keys.push(k);
  const esc = v => {
    if (v == null) return '';
    let s = (typeof v === 'object') ? JSON.stringify(v) : String(v);
    if (/^[=+\-@\t\r\n]/.test(s)) s = "'" + s;
    return /[",\n\r]/.test(s) ? '"' + s.replace(/"/g, '""') + '"' : s;
  };
  const lines = [keys.map(esc).join(',')];
  for (const r of rows) lines.push(keys.map(k => esc(r[k])).join(','));
  return '\uFEFF' + lines.join('\r\n'); // BOM：Excel 直接打开不乱码
}

/* ---------------- 入口 ---------------- */

async function main() {
  if (!CONFIG.X_TOKEN || /^(?:在此粘贴|x{8,})/i.test(CONFIG.X_TOKEN)) {
    throw new Error('请通过 ONESYSTEM_X_TOKEN 环境变量提供 sessionid（不要写入源码）');
  }
  const cmd = process.argv[2];
  if (cmd === 'calendar') { const cur = await currentTerm(); console.log('当前学期: ' + cur.id + '  ' + cur.name); for (const c of await fetchCalendars()) console.log('  ' + c.id + '  ' + c.name); return; }
  if (cmd === 'options') { await cmdOptions(); return; }

  const raw = String(CONFIG.FILTERS.calendarId || 'current');
  const cal = raw === 'current' ? (await currentTerm()).id : raw;
  console.log('抓取学期: ' + cal);
  console.log('筛选条件: ' + JSON.stringify(CONFIG.FILTERS));

  const rows = await fetchAll(cal);
  const ts = new Date().toISOString().replace(/[:T]/g, '-').slice(0, 19);
  const base = `courses-${cal}-${ts}`;
  const safeRows = sanitizeRows(rows);
  fs.writeFileSync(base + '-sanitized.json', JSON.stringify(safeRows, null, 1), 'utf8');
  fs.writeFileSync(base + '.csv', toCSV(safeRows), 'utf8');
  console.log(`完成：${safeRows.length} 条 -> ${base}-sanitized.json / ${base}.csv`);
}

main().catch(e => { console.error('错误: ' + e.message); process.exit(1); });
